logtracer/jsonlog.py

```diff
@@ -0,0 +1,116 @@
+import logging
+import sys
+import traceback
+from datetime import datetime
+
+from pythonjsonlogger import jsonlogger
+
+from logtracer import _b3, _global_vars
+
+LOG_SEVERITIES = {
+    'DEBUG': 'DEBUG',
+    'INFO': 'INFO',
+    'WARNING': 'WARNING',
+    'WARN': 'WARNING',
+    'ERROR': 'ERROR',
+    'EXCEPTION': 'ERROR'
+}
+
+
+class StackdriverJsonFormatter(jsonlogger.JsonFormatter):
+    def add_fields(self, log_record, record, message_dict):
+        """
+        Add GCP StackDriver fields (https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry) to JSON logger
+        output and remove default unused fields.
+        """
+
+        if record.exc_info:
+            exception_class, exception, tb = record.exc_info
+            tb_str = "\n".join(traceback.format_tb(tb))
+            record.message = f"""Exception: {exception_class.__name__}({str(exception)})\nTraceback:\n{tb_str}"""
+
+        gcp_log_record = {
+            'severity': LOG_SEVERITIES[record.levelname],
+            'time': datetime.fromtimestamp(record.created),
+            'logging.googleapis.com/sourceLocation': {
+                'file': record.pathname,
+                'line': record.lineno,
+                'function': record.funcName
+            },
+            'message': f'{record.name} - {record.message}'
+        }
+
+        b3_values = _b3.values()
+        if b3_values.get(_b3.B3_TRACE_ID):
+            trace_name = f'projects/{_global_vars.gcp_project_name}/traces/{b3_values[_b3.B3_TRACE_ID]}'
+            gcp_log_record.update({
+                'logging.googleapis.com/trace': trace_name,
+                'logging.googleapis.com/spanId': b3_values[_b3.B3_SPAN_ID],
+            })
+
+        log_record.update(gcp_log_record)
+        log_record.pop('exc_info', None)
+
+
+class LocalJsonFormatter(jsonlogger.JsonFormatter):
+    def add_fields(self, log_record, record, message_dict):
+        """
+        Add GCP StackDriver fields (https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry) to JSON logger
+        output and remove default unused fields.
+        """
+        if record.exc_info:
```

Comment:
Is there a way of sharing the code common between the two classes rather than repeating it?